File: Top_Interview_Questions/courseSchedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adjList = [[] for _ in range(numCourses)]
        for c1, c2 in prerequisites:
            # c1 -> c2 is an edge
            adjList[c1].append(c2)

        # state[i] = 0 (not visited) | -1 (processing) | 1 (visited/finished) 
        state = [0 for _ in range(numCourses)]

        def hasCycle(i):
            # still processing means we found cycle containing ci
            if state[i] == -1:
                return True

            # ci already visited, no need to dfs
            if state[i] == 1:
                return False 
            
            state[i] = -1
            for dep in adjList[i]:
                if hasCycle(dep):
                    return True
            state[i] = 1
            return False

        for ci in range(numCourses):
            if hasCycle(ci):
                return False # if cycle involving course, then can't finish

        return True

File: NeetCode_Roadmap/course_schedule.cpp
class Solution {
public:
    bool hasCycle(vector<vector<int>>& adj_list, vector<int>& state, int cur) {
        if (state[cur] == 1) return true;
        if (state[cur] == 0) return false;
        state[cur] = 1;
        for (auto n : adj_list[cur]) {
            if (hasCycle(adj_list, state, n)) return true;
        }
        state[cur] = 0;
        return false;
    }
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> adj_list(numCourses, vector<int>());
        for (const auto& vec : prerequisites) {
            int a = vec[0];
            int b = vec[1];
            adj_list[a].push_back(b);
        }
        for (int i = 0; i < numCourses; ++i) {
            vector<int> state(numCourses, -1);
            if (hasCycle(adj_list, state, i)) return false;
        }
        return true;
    }
};
